File: sdk/java/KamiClient.java
package com.example.kami.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * 多项目卡密授权中心 - Java SDK
 * Version: 1.0.0
 *
 * <p>使用示例:
 * <pre>
 * KamiClient client = new KamiClient(
 *     "https://api.example.com",
 *     "abc123X",
 *     "your_api_key",
 *     "your_api_secret"
 * );
 *
 * // 验证卡密
 * Map<String, Object> result = client.verifyCard("ABCD-EFGH-IJKL");
 *
 * // 消费卡密
 * result = client.consumeCard("ABCD-EFGH-IJKL", "order_001", 1, null);
 * </pre>
 */
public class KamiClient {

    private static final String CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghjkmnpqrstuvwxyz23456789";
    private static final SecureRandom RANDOM = new SecureRandom();
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final String baseUrl;
    private final String projectToken;
    private final String apiKey;
    private final String apiSecret;
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;

    public KamiClient(String baseUrl, String projectToken, String apiKey, String apiSecret) {
        this(baseUrl, projectToken, apiKey, apiSecret, 30);
    }

    public KamiClient(String baseUrl, String projectToken, String apiKey, String apiSecret, int timeoutSeconds) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.projectToken = projectToken;
        this.apiKey = apiKey;
        this.apiSecret = apiSecret;
        this.objectMapper = new ObjectMapper();
        this.httpClient = new OkHttpClient.Builder()
            .connectTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .writeTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .build();
    }

    // ========== 内部工具方法 ==========

    private String generateSignature(String apiKey, String timestamp, String nonce, String body, String apiSecret) {
        String data = apiKey + timestamp + nonce + body + apiSecret;
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(data.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hash);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate signature", e);
        }
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }

    private String generateNonce(int length) {
        StringBuilder nonce = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            nonce.append(CHARS.charAt(RANDOM.nextInt(CHARS.length())));
        }
        return nonce.toString();
    }

    private Map<String, Object> request(String endpoint, Map<String, Object> payload) {
        long timestamp = System.currentTimeMillis() / 1000;
        String nonce = generateNonce(16);
        String body = "";

        try {
            if (payload != null) {
                body = objectMapper.writeValueAsString(payload);
            }

            String signature = generateSignature(apiKey, String.valueOf(timestamp), nonce, body, apiSecret);
            String url = baseUrl + "/api/p/" + projectToken + endpoint;

            Request request = new Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("X-API-KEY", apiKey)
                .addHeader("X-TIMESTAMP", String.valueOf(timestamp))
                .addHeader("X-NONCE", nonce)
                .addHeader("X-SIGN", signature)
                .post(RequestBody.create(body, JSON))
                .build();

            try (Response response = httpClient.newCall(request).execute()) {
                String responseBody = response.body() != null ? response.body().string() : "{}";
                return objectMapper.readValue(responseBody, Map.class);
            }
        } catch (IOException e) {
            Map<String, Object> errorResult = new HashMap<>();
            errorResult.put("success", false);
            errorResult.put("code", "REQUEST_ERROR");
            errorResult.put("message", e.getMessage());
            return errorResult;
        }
    }

    // ========== 卡密相关接口 ==========

    /**
     * 生成卡密
     */
    public Map<String, Object> generateCards(int quantity, String cardType, int countValue,
                                              int durationDays, boolean isPermanent,
                                              String orderNo, Long packageId, String remark) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("quantity", quantity);
        payload.put("card_type", cardType);
        payload.put("count_value", countValue);
        payload.put("duration_days", durationDays);
        payload.put("is_permanent", isPermanent);
        if (orderNo != null) payload.put("order_no", orderNo);
        if (packageId != null) payload.put("package_id", packageId);
        if (remark != null) payload.put("remark", remark);

        return request("/cards/generate", payload);
    }

    /**
     * 验证卡密
     */
    public Map<String, Object> verifyCard(String cardKey) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("card_key", cardKey);
        return request("/cards/verify", payload);
    }

    /**
     * 查询卡密状态
     */
    public Map<String, Object> getCardStatus(String cardKey) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("card_key", cardKey);
        return request("/cards/status", payload);
    }

    /**
     * 消费卡密
     */
    public Map<String, Object> consumeCard(String cardKey, String bizId, int amount, String deviceId) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("card_key", cardKey);
        payload.put("biz_id", bizId);
        payload.put("amount", amount);
        if (deviceId != null) payload.put("device_id", deviceId);

        return request("/cards/consume", payload);
    }

    /**
     * 兑换卡密到用户权益
     */
    public Map<String, Object> redeemCard(String cardKey, String userId) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("card_key", cardKey);
        payload.put("user_id", userId);
        return request("/cards/redeem", payload);
    }

    /**
     * 解绑设备
     */
    public Map<String, Object> unbindDevice(String cardKey, String reason) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("card_key", cardKey);
        if (reason != null) payload.put("reason", reason);
        return request("/cards/unbind-device", payload);
    }

    /**
     * 退还消费
     */
    public Map<String, Object> refundConsume(String cardKey, String bizId) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("card_key", cardKey);
        payload.put("biz_id", bizId);
        return request("/cards/refund-consume", payload);
    }

    // ========== 用户权益相关接口 ==========

    /**
     * 查询用户权益状态
     */
    public Map<String, Object> getEntitlementStatus(String userId, String entitlementType) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("user_id", userId);
        payload.put("entitlement_type", entitlementType);
        return request("/entitlements/status", payload);
    }

    /**
     * 消费用户权益
     */
    public Map<String, Object> consumeEntitlement(String userId, String entitlementType,
                                                   String bizId, int amount) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("user_id", userId);
        payload.put("entitlement_type", entitlementType);
        payload.put("biz_id", bizId);
        payload.put("amount", amount);
        return request("/entitlements/consume", payload);
    }

    /**
     * 退还权益消费
     */
    public Map<String, Object> refundEntitlement(String userId, String entitlementType,
                                                  String bizId, int amount) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("user_id", userId);
        payload.put("entitlement_type", entitlementType);
        payload.put("biz_id", bizId);
        payload.put("amount", amount);
        return request("/entitlements/refund-consume", payload);
    }

    // ========== 使用示例 ==========

    public static void main(String[] args) {
        KamiClient client = new KamiClient(
            "https://api.example.com",
            "abc123X",
            "your_api_key_here",
            "your_api_secret_here"
        );

        // 示例1：生成卡密
        System.out.println("=== 生成卡密 ===");
        Map<String, Object> result = client.generateCards(
            5, "count", 100, 0, false,
            "ORDER20260621001", null, "测试批次"
        );
        System.out.println(result);

        // 示例2：验证卡密
        if (Boolean.TRUE.equals(result.get("success"))) {
            System.out.println("\n=== 验证卡密 ===");
            result = client.verifyCard("ABCD-EFGH-IJKL");
            System.out.println(result);

            // 示例3：消费卡密
            System.out.println("\n=== 消费卡密 ===");
            result = client.consumeCard("ABCD-EFGH-IJKL", "user123_001", 1, null);
            System.out.println(result);
        }
    }
}

File: sdk/python/kami_client.py
"""
多项目卡密授权中心 - Python SDK
Version: 1.0.0
"""

import hashlib
import time
import random
import string
import requests
from typing import Dict, Optional, List


class KamiClient:
    """
    卡密授权中心客户端

    使用示例:
        client = KamiClient(
            base_url="https://api.example.com",
            project_token="abc123X",
            api_key="your_api_key",
            api_secret="your_api_secret"
        )

        # 验证卡密
        result = client.verify_card("ABCD-EFGH-IJKL")

        # 消费卡密
        result = client.consume_card("ABCD-EFGH-IJKL", biz_id="order_001", amount=1)
    """

    def __init__(self, base_url: str, project_token: str, api_key: str, api_secret: str, timeout: int = 30):
        """
        初始化客户端

        :param base_url: API 基础地址，例如 https://api.example.com
        :param project_token: 7位项目访问码
        :param api_key: API Key
        :param api_secret: API Secret
        :param timeout: 请求超时时间（秒）
        """
        self.base_url = base_url.rstrip('/')
        self.project_token = project_token
        self.api_key = api_key
        self.api_secret = api_secret
        self.timeout = timeout
        self.endpoint_prefix = f"/api/p/{project_token}"

    def _generate_signature(self, timestamp: int, nonce: str, body: str) -> str:
        """生成 HMAC-SHA256 签名"""
        data = f"{self.api_key}{timestamp}{nonce}{body}{self.api_secret}"
        return hashlib.sha256(data.encode('utf-8')).hexdigest()

    def _generate_nonce(self, length: int = 16) -> str:
        """生成随机 nonce"""
        return ''.join(random.choices(string.ascii_letters + string.digits, k=length))

    def _request(self, endpoint: str, payload: Optional[Dict] = None) -> Dict:
        """
        发送 API 请求

        :param endpoint: 端点路径，例如 /cards/verify
        :param payload: 请求体（字典）
        :return: 响应 JSON
        """
        timestamp = int(time.time())
        nonce = self._generate_nonce()
        body = ""

        if payload:
            import json
            body = json.dumps(payload, separators=(',', ':'), ensure_ascii=False)

        signature = self._generate_signature(timestamp, nonce, body)

        headers = {
            "Content-Type": "application/json",
            "X-API-KEY": self.api_key,
            "X-TIMESTAMP": str(timestamp),
            "X-NONCE": nonce,
            "X-SIGN": signature
        }

        url = f"{self.base_url}{self.endpoint_prefix}{endpoint}"

        try:
            response = requests.post(url, headers=headers, json=payload, timeout=self.timeout)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {
                "success": False,
                "code": "REQUEST_ERROR",
                "message": str(e)
            }

    # ========== 卡密相关接口 ==========

    def generate_cards(self, quantity: int, card_type: str = "count",
                       count_value: int = 0, duration_days: int = 0,
                       is_permanent: bool = False, order_no: Optional[str] = None,
                       package_id: Optional[int] = None, remark: Optional[str] = None) -> Dict:
        """
        生成卡密

        :param quantity: 生成数量
        :param card_type: 卡密类型 (count/time/hybrid/permanent)
        :param count_value: 次数值
        :param duration_days: 有效天数
        :param is_permanent: 是否永久
        :param order_no: 订单号（幂等）
        :param package_id: 套餐ID
        :param remark: 备注
        :return: 生成结果
        """
        payload = {
            "quantity": quantity,
            "card_type": card_type,
            "count_value": count_value,
            "duration_days": duration_days,
            "is_permanent": is_permanent
        }
        if order_no: payload["order_no"] = order_no
        if package_id: payload["package_id"] = package_id
        if remark: payload["remark"] = remark

        return self._request("/cards/generate", payload)

    def verify_card(self, card_key: str) -> Dict:
        """
        验证卡密

        :param card_key: 卡密
        :return: 验证结果
        """
        return self._request("/cards/verify", {"card_key": card_key})

    def get_card_status(self, card_key: str) -> Dict:
        """
        查询卡密状态

        :param card_key: 卡密
        :return: 状态信息
        """
        return self._request("/cards/status", {"card_key": card_key})

    def consume_card(self, card_key: str, biz_id: str, amount: int = 1,
                     device_id: Optional[str] = None) -> Dict:
        """
        消费卡密

        :param card_key: 卡密
        :param biz_id: 业务ID（幂等）
        :param amount: 消费次数
        :param device_id: 设备ID
        :return: 消费结果
        """
        payload = {
            "card_key": card_key,
            "biz_id": biz_id,
            "amount": amount
        }
        if device_id: payload["device_id"] = device_id

        return self._request("/cards/consume", payload)

    def redeem_card(self, card_key: str, user_id: str) -> Dict:
        """
        兑换卡密到用户权益

        :param card_key: 卡密
        :param user_id: 用户ID
        :return: 兑换结果
        """
        return self._request("/cards/redeem", {
            "card_key": card_key,
            "user_id": user_id
        })

    def unbind_device(self, card_key: str, reason: Optional[str] = None) -> Dict:
        """
        解绑设备

        :param card_key: 卡密
        :param reason: 解绑原因
        :return: 解绑结果
        """
        payload = {"card_key": card_key}
        if reason: payload["reason"] = reason

        return self._request("/cards/unbind-device", payload)

    def refund_consume(self, card_key: str, biz_id: str) -> Dict:
        """
        退还消费

        :param card_key: 卡密
        :param biz_id: 原消费的业务ID
        :return: 退还结果
        """
        return self._request("/cards/refund-consume", {
            "card_key": card_key,
            "biz_id": biz_id
        })

    # ========== 用户权益相关接口 ==========

    def get_entitlement_status(self, user_id: str, entitlement_type: str) -> Dict:
        """
        查询用户权益状态

        :param user_id: 用户ID
        :param entitlement_type: 权益类型
        :return: 权益状态
        """
        return self._request("/entitlements/status", {
            "user_id": user_id,
            "entitlement_type": entitlement_type
        })

    def consume_entitlement(self, user_id: str, entitlement_type: str,
                           biz_id: str, amount: int = 1) -> Dict:
        """
        消费用户权益

        :param user_id: 用户ID
        :param entitlement_type: 权益类型
        :param biz_id: 业务ID（幂等）
        :param amount: 消费次数
        :return: 消费结果
        """
        return self._request("/entitlements/consume", {
            "user_id": user_id,
            "entitlement_type": entitlement_type,
            "biz_id": biz_id,
            "amount": amount
        })

    def refund_entitlement(self, user_id: str, entitlement_type: str,
                          biz_id: str, amount: int = 1) -> Dict:
        """
        退还权益消费

        :param user_id: 用户ID
        :param entitlement_type: 权益类型
        :param biz_id: 原消费的业务ID
        :param amount: 退还次数
        :return: 退还结果
        """
        return self._request("/entitlements/refund-consume", {
            "user_id": user_id,
            "entitlement_type": entitlement_type,
            "biz_id": biz_id,
            "amount": amount
        })


# ========== 使用示例 ==========

if __name__ == "__main__":
    # 初始化客户端
    client = KamiClient(
        base_url="https://api.example.com",
        project_token="abc123X",
        api_key="your_api_key_here",
        api_secret="your_api_secret_here"
    )

    # 示例1：生成卡密
    print("=== 生成卡密 ===")
    result = client.generate_cards(
        quantity=5,
        card_type="count",
        count_value=100,
        order_no="ORDER20260621001"
    )
    print(result)

    # 示例2：验证卡密
    if result.get("success") and result.get("cards"):
        card_key = result["cards"][0]["card_key"]
        print(f"\n=== 验证卡密: {card_key} ===")
        result = client.verify_card(card_key)
        print(result)

        # 示例3：消费卡密
        print(f"\n=== 消费卡密 ===")
        result = client.consume_card(
            card_key=card_key,
            biz_id="user123_20260621_001",
            amount=1
        )
        print(result)

        # 示例4：查询状态
        print(f"\n=== 查询状态 ===")
        result = client.get_card_status(card_key)
        print(result)
